### PR TITLE
UCT/IB/DC: Unite DC to one object

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8400/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8401/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@brminich

File: src/uct/ib/dc/dc_mlx5.c [hunk 1]
```diff
@@ -1393,61 +1416,596 @@ static ucs_status_t uct_dc_mlx5_iface_tag_init(uct_dc_mlx5_iface_t *iface,
...
+/* take dc qp to rts state */
+ucs_status_t uct_dc_mlx5_iface_dci_connect(uct_dc_mlx5_iface_t *iface,
+                                           uct_rc_txqp_t *dci)
+{
+    struct ibv_exp_qp_attr attr;
+    long attr_mask;
+
+    memset(&attr, 0, sizeof(attr));
+    attr.qp_state        = IBV_QPS_INIT;
+    attr.pkey_index      = 0;
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c]: minor: we just set it to `0` by `memset`

File: src/uct/ib/dc/dc_mlx5.c [hunk 2]
```diff
@@ -1393,61 +1416,596 @@ static ucs_status_t uct_dc_mlx5_iface_tag_init(uct_dc_mlx5_iface_t *iface,
...
+    attr.path_mtu                   = iface->super.config.path_mtu;
+    attr.min_rnr_timer              = 0;
+    attr.max_dest_rd_atomic         = 1;
+    attr.ah_attr.is_global          = iface->super.super.is_global_addr;
+    attr.ah_attr.sl                 = iface->super.super.config.sl;
+    attr_mask                       = IBV_EXP_QP_STATE     |
+                                      IBV_EXP_QP_PATH_MTU  |
+                                      IBV_EXP_QP_AV;
+
+#if HAVE_DECL_IBV_EXP_QP_OOO_RW_DATA_PLACEMENT
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c]: you could create an inline function for this, as it is used twice at least

File: src/uct/ib/dc/dc_mlx5.c [hunk 3]
```diff
@@ -1393,61 +1416,596 @@ static ucs_status_t uct_dc_mlx5_iface_tag_init(uct_dc_mlx5_iface_t *iface,
...
+
+void uct_dc_destroy_dct(uct_dc_mlx5_iface_t *iface)
+{
+    if (iface->rx_dct != NULL) {
+        ibv_exp_destroy_dct(iface->rx_dct);
+    }
+    iface->rx_dct = NULL;
+}
+
+static ucs_status_t uct_dc_device_init(uct_ib_device_t *dev)
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c]: `uct_dc_mlx5_device_init` ?

File: src/uct/ib/dc/dc_mlx5.c [hunk 4]
```diff
@@ -1393,61 +1416,596 @@ static ucs_status_t uct_dc_mlx5_iface_tag_init(uct_dc_mlx5_iface_t *iface,
...
+    uct_dc_mlx5_ep_pending_purge(&iface->tx.fc_ep->super.super, NULL, NULL);
+    ucs_arbiter_group_cleanup(&iface->tx.fc_ep->arb_group);
+    uct_rc_fc_cleanup(&iface->tx.fc_ep->fc);
+    UCS_CLASS_CLEANUP(uct_base_ep_t, iface->tx.fc_ep);
+    ucs_free(iface->tx.fc_ep);
+}
+
+ucs_status_t uct_dc_mlx5_iface_fc_grant(uct_pending_req_t *self)
+{
+    ucs_status_t status;
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c]: plz move it after initialized vars

File: src/uct/ib/dc/dc_mlx5.c [hunk 5]
```diff
@@ -1393,61 +1416,596 @@ static ucs_status_t uct_dc_mlx5_iface_tag_init(uct_dc_mlx5_iface_t *iface,
...
+
+ucs_status_t uct_dc_mlx5_iface_fc_handler(uct_rc_iface_t *rc_iface, unsigned qp_num,
+                                          uct_rc_hdr_t *hdr, unsigned length,
+                                          uint32_t imm_data, uint16_t lid, unsigned flags)
+{
+    uct_dc_mlx5_ep_t    *ep;
+    ucs_status_t        status;
+    int16_t             cur_wnd;
+    uct_dc_fc_request_t *dc_req;
+    uint8_t             fc_hdr = uct_rc_fc_get_fc_hdr(hdr->am_id);
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c]: plz move initialized vars before others

File: src/uct/ib/dc/dc_mlx5.c [hunk 6]
```diff
@@ -1511,12 +2085,11 @@ ucs_status_t uct_dc_mlx5_query_resources(uct_md_h md,
                                             resources_p, num_resources_p);
 }
 
-
 UCT_TL_COMPONENT_DEFINE(uct_dc_mlx5_tl,
                         uct_dc_mlx5_query_resources,
                         uct_dc_mlx5_iface_t,
                         "dc_mlx5",
-                        "DC_MLX5_",
+                        "DC_",
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c]: Why `DC`? I guess we should continue using `...DC_MLX5...` config names

File: src/uct/ib/dc/dc_mlx5.h [hunk 7]
```diff
@@ -27,14 +30,14 @@ typedef enum {
     UCT_DC_IFACE_ADDR_DC_V2   = UCS_BIT(2),
     UCT_DC_IFACE_ADDR_DC_VERS = UCT_DC_IFACE_ADDR_DC_V1 |
                                 UCT_DC_IFACE_ADDR_DC_V2
-} uct_dc_iface_addr_flags_t;
+} uct_dc_mlx5_iface_addr_flags_t;
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.h]: worth to include `MLX5` to flag names

File: src/uct/ib/dc/dc_mlx5.h [hunk 8]
```diff
@@ -111,74 +111,85 @@ struct uct_dc_iface {
...
+#elif HAVE_DC_DV
+    struct ibv_qp                 *rx_dct;
 #endif
 
     uint8_t                       version_flag;
-};
 
+    uct_rc_mlx5_iface_common_t    mlx5_common;
+    uct_ud_mlx5_iface_common_t    ud_common;
+    uct_ib_mlx5_txwq_t            dci_wqs[UCT_DC_IFACE_MAX_DCIS];
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.h]: I'd move it to `tx` part of this struct

File: src/uct/ib/dc/dc_mlx5.h [hunk 9]
```diff
@@ -111,74 +111,85 @@ struct uct_dc_iface {
...
 
-int uct_dc_iface_is_reachable(const uct_iface_h tl_iface,
-                              const uct_device_addr_t *dev_addr,
-                              const uct_iface_addr_t *iface_addr);
+ucs_status_t uct_dc_mlx5_iface_create_dct(uct_dc_mlx5_iface_t *iface);
 
-ucs_status_t uct_dc_iface_get_address(uct_iface_h tl_iface, uct_iface_addr_t *iface_addr);
+int uct_dc_mlx5_iface_is_reachable(const uct_iface_h tl_iface,
+                                   const uct_device_addr_t *dev_addr,
+                                   const uct_iface_addr_t *iface_addr);
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.h]: please add line between declarations

File: src/uct/ib/dc/dc_mlx5.c [hunk 10]
```diff
@@ -5,83 +5,71 @@
...
-  {"", "", NULL,
-   ucs_offsetof(uct_dc_mlx5_iface_config_t, ud_common),
-   UCS_CONFIG_TYPE_TABLE(uct_ud_mlx5_iface_common_config_table)},
-
-  {"", "", NULL,
-   ucs_offsetof(uct_dc_mlx5_iface_config_t, mlx5_common),
-   UCS_CONFIG_TYPE_TABLE(uct_ib_mlx5_iface_config_table)},
-
-  {NULL}
+ucs_config_field_t uct_dc_mlx5_iface_config_sub_table[] = {
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/dc/dc_mlx5.c]: To keep existing parameters scheme. This way DC parameters still can be accessed with DC_ prefix, like DC_RC_IB_TX_QUEUE_LEN in test_peer_failure.cc

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8408/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@brminich

File: src/uct/ib/dc/dc_mlx5.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (brminich) [reply src/uct/ib/dc/dc_mlx5.c]: Right, probably we have to keep pure '_DC_' var names. 
@yosefe, wdyt?

[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5.c]: agree, let's keep old var names for now, and add comment to the code this is the reason for subtable

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8417/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8420/ for details (Mellanox internal link).

**Pull request merged**